### PR TITLE
Traceql better pooling

### PR DESCRIPTION
**Pull request by mdisibio**
**What this PR does**:
This PR splits up the pooling for differently sized `parquetquery.IteratorResults` which improves memory for (primarily) TraceQL metrics queries.  To summarize sometimes an IteratorResult could be a spanset with 1000's of spans attached to it, or as small as an attribute with key+value.  Because there was 1 pool, all approached the max size needed which led to over allocation.  Memory improvement gets better as the query processes more spans, which is why regular searches that exit early only improve a little.

Left and regular JoinIterators can now be given a custom pool, and for now there are 4 static pools in vParquet3:  spans, spansets, traces, and attributes.  Long-term we could look at allocating these within the execution so they are truly optimal for the given query and tenant.   Should address Unions but decided to wrap this PR up without.

Note: default pooling remains for everything else. 

Benchmarks:

<details>
<summary>BackendBlockQueryRange</summary>

```
goos: darwin
goarch: arm64
pkg: github.com/grafana/tempo/tempodb/encoding/vparquet3
                                                                             │ before.txt  │             after.txt              │
                                                                             │   sec/op    │   sec/op     vs base               │
BackendBlockQueryRange/{}_|_rate()/5-12                                        496.3m ± 2%   501.2m ± 3%       ~ (p=0.393 n=10)
BackendBlockQueryRange/{}_|_rate()_by_(name)/5-12                               1.154 ± 7%    1.149 ± 1%       ~ (p=0.481 n=10)
BackendBlockQueryRange/{}_|_rate()_by_(resource.service.name)/5-12             954.9m ± 8%   898.5m ± 3%  -5.91% (p=0.007 n=10)
BackendBlockQueryRange/{resource.service.name=`tempo-gateway`}_|_rate()/5-12   65.62m ± 1%   64.94m ± 1%  -1.04% (p=0.002 n=10)
BackendBlockQueryRange/{status=error}_|_rate()/5-12                            151.9m ± 4%   148.8m ± 0%  -2.05% (p=0.000 n=10)
geomean                                                                        352.6m        346.6m       -1.71%

                                                                             │ before.txt │              after.txt              │
                                                                             │  MB_IO/op  │  MB_IO/op   vs base                 │
BackendBlockQueryRange/{}_|_rate()/5-12                                        43.45 ± 0%   43.45 ± 0%       ~ (p=1.000 n=10) ¹
BackendBlockQueryRange/{}_|_rate()_by_(name)/5-12                              57.75 ± 0%   57.75 ± 0%       ~ (p=1.000 n=10) ¹
BackendBlockQueryRange/{}_|_rate()_by_(resource.service.name)/5-12             44.23 ± 0%   44.23 ± 0%       ~ (p=1.000 n=10) ¹
BackendBlockQueryRange/{resource.service.name=`tempo-gateway`}_|_rate()/5-12   41.85 ± 0%   41.85 ± 0%       ~ (p=1.000 n=10) ¹
BackendBlockQueryRange/{status=error}_|_rate()/5-12                            46.66 ± 0%   46.66 ± 0%       ~ (p=1.000 n=10) ¹
geomean                                                                        46.47        46.47       +0.00%
¹ all samples are equal

                                                                             │ before.txt  │              after.txt               │
                                                                             │  spans/op   │  spans/op    vs base                 │
BackendBlockQueryRange/{}_|_rate()/5-12                                        2.266M ± 0%   2.266M ± 0%       ~ (p=1.000 n=10) ¹
BackendBlockQueryRange/{}_|_rate()_by_(name)/5-12                              2.266M ± 0%   2.266M ± 0%       ~ (p=1.000 n=10) ¹
BackendBlockQueryRange/{}_|_rate()_by_(resource.service.name)/5-12             2.266M ± 0%   2.266M ± 0%       ~ (p=1.000 n=10) ¹
BackendBlockQueryRange/{resource.service.name=`tempo-gateway`}_|_rate()/5-12   1.800k ± 0%   1.800k ± 0%       ~ (p=1.000 n=10) ¹
BackendBlockQueryRange/{status=error}_|_rate()/5-12                            38.82k ± 0%   38.82k ± 0%       ~ (p=1.000 n=10) ¹
geomean                                                                        241.0k        241.0k       +0.00%
¹ all samples are equal

                                                                             │ before.txt  │             after.txt              │
                                                                             │   spans/s   │   spans/s    vs base               │
BackendBlockQueryRange/{}_|_rate()/5-12                                        4.566M ± 2%   4.521M ± 3%       ~ (p=0.393 n=10)
BackendBlockQueryRange/{}_|_rate()_by_(name)/5-12                              1.964M ± 6%   1.972M ± 1%       ~ (p=0.481 n=10)
BackendBlockQueryRange/{}_|_rate()_by_(resource.service.name)/5-12             2.373M ± 7%   2.522M ± 3%  +6.27% (p=0.007 n=10)
BackendBlockQueryRange/{resource.service.name=`tempo-gateway`}_|_rate()/5-12   27.43k ± 1%   27.72k ± 1%  +1.05% (p=0.002 n=10)
BackendBlockQueryRange/{status=error}_|_rate()/5-12                            255.6k ± 4%   260.9k ± 0%  +2.09% (p=0.000 n=10)
geomean                                                                        683.5k        695.4k       +1.74%

                                                                             │  before.txt   │               after.txt               │
                                                                             │     B/op      │     B/op       vs base                │
BackendBlockQueryRange/{}_|_rate()/5-12                                        90.10Mi ± 21%   24.42Mi ± 97%  -72.90% (p=0.000 n=10)
BackendBlockQueryRange/{}_|_rate()_by_(name)/5-12                              229.3Mi ±  8%   153.3Mi ± 51%  -33.16% (p=0.023 n=10)
BackendBlockQueryRange/{}_|_rate()_by_(resource.service.name)/5-12             78.56Mi ±  7%   48.12Mi ± 40%  -38.74% (p=0.000 n=10)
BackendBlockQueryRange/{resource.service.name=`tempo-gateway`}_|_rate()/5-12   27.08Mi ±  6%   25.40Mi ±  5%        ~ (p=0.143 n=10)
BackendBlockQueryRange/{status=error}_|_rate()/5-12                            23.78Mi ±  8%   24.86Mi ±  6%        ~ (p=0.315 n=10)
geomean                                                                        63.66Mi         40.85Mi        -35.83%

                                                                             │  before.txt  │              after.txt              │
                                                                             │  allocs/op   │  allocs/op    vs base               │
BackendBlockQueryRange/{}_|_rate()/5-12                                        358.8k ± 10%   330.9k ± 10%  -7.76% (p=0.008 n=10)
BackendBlockQueryRange/{}_|_rate()_by_(name)/5-12                              3.995M ±  2%   3.931M ±  4%       ~ (p=0.247 n=10)
BackendBlockQueryRange/{}_|_rate()_by_(resource.service.name)/5-12             719.3k ±  1%   719.2k ±  5%       ~ (p=0.565 n=10)
BackendBlockQueryRange/{resource.service.name=`tempo-gateway`}_|_rate()/5-12   298.2k ±  0%   298.2k ±  0%       ~ (p=0.724 n=10)
BackendBlockQueryRange/{status=error}_|_rate()/5-12                            306.3k ±  0%   306.3k ±  0%       ~ (p=0.565 n=10)
geomean                                                                        623.4k         611.4k        -1.92%
```
</details>

<details>
<summary>BackendBlockTraceQL</summary>

```
goos: darwin
goarch: arm64
pkg: github.com/grafana/tempo/tempodb/encoding/vparquet3
                                                 │ before.traceql.txt │         after.traceql.txt          │
                                                 │       sec/op       │   sec/op     vs base               │
BackendBlockTraceQL/spanAttValNoMatch-12                 18.03m ±  1%   17.51m ± 1%  -2.89% (p=0.000 n=10)
BackendBlockTraceQL/spanAttIntrinsicNoMatch-12           17.24m ±  0%   16.82m ± 1%  -2.41% (p=0.002 n=10)
BackendBlockTraceQL/resourceAttValNoMatch-12             17.28m ±  1%   16.98m ± 3%  -1.76% (p=0.043 n=10)
BackendBlockTraceQL/resourceAttIntrinsicMatch-12         27.73m ±  1%   27.10m ± 0%  -2.27% (p=0.000 n=10)
BackendBlockTraceQL/mixedValNoMatch-12                   148.6m ±  1%   150.7m ± 1%  +1.41% (p=0.000 n=10)
BackendBlockTraceQL/mixedValMixedMatchAnd-12             17.44m ±  1%   17.14m ± 0%  -1.71% (p=0.000 n=10)
BackendBlockTraceQL/mixedValMixedMatchOr-12              156.7m ±  0%   158.1m ± 1%  +0.90% (p=0.000 n=10)
BackendBlockTraceQL/count-12                             235.2m ±  8%   236.6m ± 0%       ~ (p=0.796 n=10)
BackendBlockTraceQL/struct-12                            415.5m ± 15%   417.5m ± 1%       ~ (p=0.796 n=10)
BackendBlockTraceQL/||-12                                138.6m ±  1%   137.4m ± 0%  -0.87% (p=0.000 n=10)
BackendBlockTraceQL/mixed-12                             39.38m ±  1%   38.78m ± 1%  -1.52% (p=0.001 n=10)
geomean                                                  59.36m         58.81m       -0.92%

                                                 │ before.traceql.txt │          after.traceql.txt          │
                                                 │        B/s         │     B/s       vs base               │
BackendBlockTraceQL/spanAttValNoMatch-12                267.7Mi ±  1%   275.7Mi ± 1%  +2.98% (p=0.000 n=10)
BackendBlockTraceQL/spanAttIntrinsicNoMatch-12          268.7Mi ±  0%   275.3Mi ± 1%  +2.47% (p=0.002 n=10)
BackendBlockTraceQL/resourceAttValNoMatch-12            210.6Mi ±  1%   214.4Mi ± 3%  +1.79% (p=0.043 n=10)
BackendBlockTraceQL/resourceAttIntrinsicMatch-12        609.3Mi ±  1%   623.4Mi ± 0%  +2.32% (p=0.000 n=10)
BackendBlockTraceQL/mixedValNoMatch-12                  34.13Mi ±  1%   33.66Mi ± 1%  -1.38% (p=0.000 n=10)
BackendBlockTraceQL/mixedValMixedMatchAnd-12            209.2Mi ±  1%   212.8Mi ± 0%  +1.74% (p=0.000 n=10)
BackendBlockTraceQL/mixedValMixedMatchOr-12             82.47Mi ±  0%   81.74Mi ± 1%  -0.88% (p=0.000 n=10)
BackendBlockTraceQL/count-12                            77.03Mi ±  7%   76.59Mi ± 0%       ~ (p=0.781 n=10)
BackendBlockTraceQL/struct-12                           35.01Mi ± 13%   34.84Mi ± 1%       ~ (p=0.853 n=10)
BackendBlockTraceQL/||-12                               133.0Mi ±  1%   134.2Mi ± 0%  +0.88% (p=0.000 n=10)
BackendBlockTraceQL/mixed-12                            428.6Mi ±  1%   435.2Mi ± 1%  +1.54% (p=0.001 n=10)
geomean                                                 149.5Mi         150.9Mi       +0.93%

                                                 │ before.traceql.txt │          after.traceql.txt          │
                                                 │      MB_io/op      │  MB_io/op   vs base                 │
BackendBlockTraceQL/spanAttValNoMatch-12                   5.061 ± 0%   5.061 ± 0%       ~ (p=1.000 n=10) ¹
BackendBlockTraceQL/spanAttIntrinsicNoMatch-12             4.856 ± 0%   4.856 ± 0%       ~ (p=1.000 n=10) ¹
BackendBlockTraceQL/resourceAttValNoMatch-12               3.817 ± 0%   3.817 ± 0%       ~ (p=1.000 n=10) ¹
BackendBlockTraceQL/resourceAttIntrinsicMatch-12           17.72 ± 0%   17.72 ± 0%       ~ (p=1.000 n=10) ¹
BackendBlockTraceQL/mixedValNoMatch-12                     5.318 ± 0%   5.318 ± 0%       ~ (p=1.000 n=10) ¹
BackendBlockTraceQL/mixedValMixedMatchAnd-12               3.825 ± 0%   3.825 ± 0%       ~ (p=1.000 n=10) ¹
BackendBlockTraceQL/mixedValMixedMatchOr-12                13.55 ± 0%   13.55 ± 0%       ~ (p=1.000 n=10) ¹
BackendBlockTraceQL/count-12                               19.00 ± 0%   19.00 ± 0%       ~ (p=1.000 n=10) ¹
BackendBlockTraceQL/struct-12                              15.25 ± 0%   15.25 ± 0%       ~ (p=1.000 n=10) ¹
BackendBlockTraceQL/||-12                                  19.33 ± 0%   19.33 ± 0%       ~ (p=1.000 n=10) ¹
BackendBlockTraceQL/mixed-12                               17.70 ± 0%   17.70 ± 0%       ~ (p=1.000 n=10) ¹
geomean                                                    9.308        9.308       +0.00%
¹ all samples are equal

                                                 │ before.traceql.txt │           after.traceql.txt           │
                                                 │        B/op        │     B/op       vs base                │
BackendBlockTraceQL/spanAttValNoMatch-12                25.69Mi ±  6%   26.45Mi ±  7%        ~ (p=0.280 n=10)
BackendBlockTraceQL/spanAttIntrinsicNoMatch-12          25.22Mi ±  0%   25.22Mi ±  0%   +0.01% (p=0.000 n=10)
BackendBlockTraceQL/resourceAttValNoMatch-12            25.07Mi ±  0%   25.07Mi ±  0%   +0.01% (p=0.000 n=10)
BackendBlockTraceQL/resourceAttIntrinsicMatch-12        20.14Mi ±  2%   20.27Mi ±  3%        ~ (p=0.971 n=10)
BackendBlockTraceQL/mixedValNoMatch-12                  22.18Mi ± 14%   22.14Mi ± 15%        ~ (p=0.853 n=10)
BackendBlockTraceQL/mixedValMixedMatchAnd-12            25.09Mi ±  0%   25.09Mi ±  0%   +0.01% (p=0.000 n=10)
BackendBlockTraceQL/mixedValMixedMatchOr-12             21.23Mi ±  5%   21.67Mi ±  6%        ~ (p=0.247 n=10)
BackendBlockTraceQL/count-12                            283.5Mi ±  1%   286.5Mi ±  2%        ~ (p=0.089 n=10)
BackendBlockTraceQL/struct-12                           54.89Mi ±  4%   44.29Mi ± 16%  -19.32% (p=0.002 n=10)
BackendBlockTraceQL/||-12                               125.6Mi ±  0%   125.0Mi ±  0%   -0.49% (p=0.000 n=10)
BackendBlockTraceQL/mixed-12                            20.56Mi ±  4%   20.59Mi ±  3%        ~ (p=0.853 n=10)
geomean                                                 36.55Mi         36.04Mi         -1.38%

                                                 │ before.traceql.txt │         after.traceql.txt          │
                                                 │     allocs/op      │  allocs/op   vs base               │
BackendBlockTraceQL/spanAttValNoMatch-12                  289.2k ± 0%   289.2k ± 0%  +0.01% (p=0.000 n=10)
BackendBlockTraceQL/spanAttIntrinsicNoMatch-12            289.1k ± 0%   289.1k ± 0%  +0.01% (p=0.000 n=10)
BackendBlockTraceQL/resourceAttValNoMatch-12              289.1k ± 0%   289.2k ± 0%  +0.01% (p=0.000 n=10)
BackendBlockTraceQL/resourceAttIntrinsicMatch-12          290.6k ± 0%   290.7k ± 0%  +0.01% (p=0.000 n=10)
BackendBlockTraceQL/mixedValNoMatch-12                    290.6k ± 0%   290.6k ± 0%       ~ (p=0.239 n=10)
BackendBlockTraceQL/mixedValMixedMatchAnd-12              289.2k ± 0%   289.2k ± 0%  +0.01% (p=0.000 n=10)
BackendBlockTraceQL/mixedValMixedMatchOr-12               290.6k ± 0%   290.6k ± 0%  +0.01% (p=0.029 n=10)
BackendBlockTraceQL/count-12                              1.706M ± 0%   1.707M ± 0%  +0.02% (p=0.000 n=10)
BackendBlockTraceQL/struct-12                             455.2k ± 0%   455.5k ± 0%       ~ (p=0.616 n=10)
BackendBlockTraceQL/||-12                                 937.6k ± 0%   937.6k ± 0%  +0.00% (p=0.000 n=10)
BackendBlockTraceQL/mixed-12                              291.9k ± 0%   291.9k ± 0%  +0.01% (p=0.000 n=10)
geomean                                                   395.0k        395.0k       +0.01%
```
</details>

**Which issue(s) this PR fixes**:
Fixes n/a

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: tempodb/encoding/vparquet3/block_traceql.go [hunk 1]
```diff
@@ -1314,7 +1321,12 @@ func createSpanIterator(makeIter makeIterFn, primaryIter parquetquery.Iterator,
 
 	// Left join here means the span id/start/end iterators + 1 are required,
 	// and all other conditions are optional. Whatever matches is returned.
-	return parquetquery.NewLeftJoinIterator(DefinitionLevelResourceSpansILSSpan, required, iters, spanCol)
+	j, err := parquetquery.NewLeftJoinIterator(DefinitionLevelResourceSpansILSSpan, required, iters, spanCol)
+	if err != nil {
+		return nil, err
+	}
+	j.Pool = pqSpanPool.Get
```
Comment (mapno) [inline tempodb/encoding/vparquet3/block_traceql.go:1328]: Not a huge fan of this, feels a bit unsafe. But IDK if another approach is worth it.

**Review by mapno — APPROVED** [review]
LGTM

File: tempodb/encoding/vparquet3/block_traceql.go (hunk 1 above)
Comment (mdisibio) [reply tempodb/encoding/vparquet3/block_traceql.go]: 👍 Switched to options pattern.

**Review by mapno — APPROVED** [review]
Nice

**Pull request merged**